fendermint/vm/interpreter/src/chain.rs

```diff
@@ -177,32 +217,87 @@ where
-                    Ok(((pool, state), ChainMessageApplyRet::Signed(ret)))
+                    Ok(((pool, provider, state), ChainMessageApplyRet::Signed(ret)))
                 }
                 IpcMessage::BottomUpExec(_) => {
                     todo!("#197: implement BottomUp checkpoint execution")
                 }
-                IpcMessage::TopDown => {
-                    todo!("implement TopDown handling; this is just a placeholder")
+                IpcMessage::TopDownExec(p) => {
+                    if !provider.is_enabled() {
+                        return Err(anyhow!(
+                            "cannot execute IPC top-down message: parent provider disabled"
+                        ));
+                    }
+
+                    // commit parent finality first
+                    let finality = IPCParentFinality {
+                        height: p.height as u64,
+                        block_hash: p.block_hash,
+                    };
+                    let msg = encode_to_fvm_implicit(&encode_commit_parent_finality_call(
+                        finality.clone(),
+                    )?)?;
+                    let (state, ret) = self
+                        .inner
+                        .deliver(state, VerifiableMessage::NotVerify(msg))
+                        .await?;
+                    if ret.is_err() {
+                        // TODO: how to handle error here? Do we return Err or stash them into one ret?
+                        todo!()
+                    }
+
+                    // stash validator changes
+
+                    // error happens if we cannot get the validator set from ipc agent after retries
+                    let validator_changes = provider.validator_changes(p.height as u64).await?;
+                    let msg = encode_to_fvm_implicit(&encode_store_validator_changes_call(
+                        validator_changes,
+                    )?)?;
+                    let (state, ret) = self
+                        .inner
+                        .deliver(state, VerifiableMessage::NotVerify(msg))
+                        .await?;
+                    if ret.is_err() {
+                        // TODO: how to handle error here? Do we return Err or stash them into one ret?
+                        todo!()
+                    }
+
+                    // Execute top down messages
+
+                    // error happens if we cannot get the validator set from ipc agent after retries
+                    let messages = provider.top_down_msgs(p.height as u64).await?;
+                    let msg = encode_to_fvm_implicit(&encode_apply_cross_messages_call(messages)?)?;
+                    let (state, ret) = self
+                        .inner
+                        .deliver(state, VerifiableMessage::NotVerify(msg))
+                        .await?;
+                    if ret.is_err() {
+                        // TODO: how to handle error here? Do we return Err or stash them into one ret?
```

Comment:
Once again, crashing might be our best option to avoid invalid state. 

I agree that the messages in the batch should be executed separately and their result should be emitted as events regardless of the outcome - at least I can't think of a better way of surfacing them. 

But the overall execution itself should not fail. If there is a reason it fails, we should fix it and resume from there, rather than sometimes skip entire batches because of a rotten apple or who knows what causes it to fail.

Reply:
For example say it fails because even implicit executions respect the gas limit, which we set at `BLOCK_GAS_LIMIT`. Even though we talked about it when I proposed that messages should be appended to a queue, not immediately executed, and then executed as and when there is unused gas in the block, at this point it's just one big bang and it can fail. 

What then, do we just say oh well, sometimes depending on how many messages there are we skip entire batches? Or should we say we quickly patch the application to try to break it up into multiple batches? 

Of course crashing all nodes probably means the chain outage is going to be unacceptable, which leads me to back to my point that the best thing to do would be to treat the entire top-down execution as an atomic unit and mark it as failed, so that it can be re-tried again and again until it's fixed, but without crashing Fendermint or halting all other applications. 

Sorry, I know I am going back and forth from crashing to marking as failed, but in this case there is a legitimate reason to for failure, which is this overall gas limit. What we should not do is partial failure IMO.

Reply:
Another relatively easy to implement option to prevent running out of gas while handling the batches is to do https://github.com/consensus-shipyard/fendermint/issues/261